### ISSUE TITLE
Ray on Kubeflow - Asking for user and contributor interest / commentss

### ISSUE THREAD

**Issue by jbottum** [kubeflow/kubeflow#6680]
/kind feature

**Why you need this feature:**
Ray provides a preprocessing, training, tuning and serving framework and a growing number of data scientist are using it i.e. the Ray Community has nearly 5,000 slack members.   Ray provides these advanced libraries:   

- RLLib for reinforcement learning
- Ray Tune for hyperparameter tuning
- Ray Train for distributed deep learning
- Ray Serve for scalable model serving
- Ray Data for preprocessing

At the Kubeflow Summit, several Kubeflow members expressed interest in Richard Liu's presentation on Ray on Kubeflow, which is attached.   [Ray with Kubeflow.pdf](https://github.com/kubeflow/kubeflow/files/9832933/Ray.with.Kubeflow.pdf)

The solution is also explained in this post: https://cloud.google.com/blog/products/ai-machine-learning/build-a-ml-platform-with-kubeflow-and-ray-on-gke

**Describe the solution you'd like:**
[A clear and concise description of what you want to happen.]

To quote Richard Liu, "What if we can combine the benefits of both systems (Ray and Kubeflow) ? Imagine having an environment that:

- Supports Ray Train with autoscaling and resource provisioning
- Integrated with identity-based authentication and authorization
- Supports multi-user isolation and collaboration
- Contains an interactive notebook server"

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

Enabling a phase 1 design of Ray support on Kubeflow seems to be a good goal for Kubeflow v1.7 (March '23).   To move this forward, it would be great if users would express interest.   If enough interest is generated, then we could ask some SME experts to create a design doc or PR for a technical review.   

**At this stage, it would be great if interested users and/or contributors would provide their interest and comments below (so that we can review them in the Oct 25 Kubeflow Community Meeting). **   

Thanks to @richardsliu @Jeffwan and Winston Chiang.    and adding for visibility @johnugeorge @kimwnasptd @DomFleischmann

**Comment by richardliaw** [comment]
Hey there, this is great! Reaching out from the Ray team. If you're interested in collaborating on this, let's chat.

cc @DmitriGekhtman who should know more about similar user requests here.

**Comment by DmitriGekhtman** [comment]
Hi!

I'd like to link the corresponding issue on the Ray side:
https://github.com/ray-project/kuberay/issues/502

I'll follow up with more thoughts soon.

**Comment by goswamig** [comment]
+1 for this feature.

**Comment by PierrickPochelu** [comment]
I would love contribute to this. I use Ray since 2019. AFAK it is the best tool combining performance+completeness for MLOps.

**Comment by mjw2194** [comment]
I would strongly advocate for this. It seems to me that Ray would provide a baseline for distributed computing capabilities, giving users much more extensibility.

**Comment by juliusvonkohout** [comment]
I have already started a few days ago with the architecture.
There are some limitations that need to be mitigated so satisfy the new /contrib requirements merged a few days ago.

1. Enforce Runasnonroot for all images: Remove the root rights from the rayworker according to https://github.com/ray-project/ray/issues/14077#issuecomment-1283208595. I think udocker might be most promising. Most other images should already be compatible.  @richardliaw can you take care of that?
2. Namespace isolation: We need proper istio authorizationpolicies (profile namespace) and RBAC aggregation rules for default-editor/jupyter-web-app to manage clusters. a simple thing for me
3. We need to integrate ray into the dashboard, since we deploy at least one cluster per namespace. that should be easy
4. ray-system needs istio authorizationpolicies and networkpolicies. also something i can do

The most interesting thing will be the automation approach such that users can create and delete rayclusters on demand similar to inferenceservices, jupyterlab, rstudio,vscode, labelstudio and others. So we need something similar to the kserve/katib web application and controller to mange ray clusters.  Managing(create/delete, choosing name, image and resources) and showing/proxying the corresponding ray cluster dashboard should be fine for now.

The user must be able to configure 
clustername
head and worker images
head and worker resources
          resources:
            limits:
              cpu: "1"
              memory: "8G"
            requests:
              cpu: "500m"
              memory: "4G"
worker replicas
  - replicas: 3
    minReplicas: 1
    maxReplicas: 10

via the GUI. Editing already existing clusters is not necessary at the moment. Due to proper autoscaling i think starting and stopping is not needed yet.

I can provide architectural guidance to volunteers, and handle most of the stuff, but someone needs to write the corresponding controller and web application. @PierrickPochelu  can you handle that?

**Comment by juliusvonkohout** [comment]
@DmitriGekhtman do you have another architecture in mind, that aligns with the kubeflow design patterns and requirements ?

**Comment by juliusvonkohout** [comment]
> 

We can discuss this on slack. Maybe have a group and meeting to discuss it.

**Comment by BACtaki** [comment]
This is really interesting! Would love to be part of the discussion

**Comment by jbottum** [comment]
All - I have added this issue to the Kubeflow Community Call Agenda for Oct 25 at 8am PST.   We will probably get to it about 8:20am.  I expect that we will spend ~10 minutes to review.   Does anyone want to help lead or co-lead this initial discussion to the Community?

Bridge: https://zoom.us/j/83469905816?pwd=aWM5clpLOTNTMU9udFpkZmV0L01VQT09

Meeting Notes: https://docs.google.com/document/d/1Wdxt1xedAj7qF_Rjmxy1R0NRdfv7UWs-r2PItewxHpE/edit#

**Comment by juliusvonkohout** [comment]
@jbottum i can but have to leave early after 20 Minutes. 
@BACtaki just write me on slack.

**Comment by DmitriGekhtman** [comment]
@juliusvonkohout

Re: Configuration of RayCluster in UI. 

I believe this is what the KubeRay API server was built to support https://ray-project.github.io/kuberay/components/apiserver/. It's an option to use that here.
cc @Jeffwan 

Re: Enforcing runasnonroot for workers.

Can you elaborate on this a bit more? Is this is an issue for the Ray team to solve? Can we modify or provide more options for how we build docker images?

Re: Architecture in general

I am not at all a KubeFlow expert, but from the Ray side one question/concern is the mechanism used to connect from the notebook to the Ray cluster. Right now the proposal is to use Ray Client. The trouble with this is that the Ray team considers Ray Client to be a functionality in maintenance-mode and wishes to emphasize other modes of interacting with Ray clusters. Some alternatives to Ray client:

- It may be advantageous to simply start a Ray worker node in the notebook server's pod -- however, for this to work well, the Ray team would need to provide a clean way of creating Ray nodes which do not execute workload processes.  

- Use Ray Job submission: 
https://docs.ray.io/en/latest/cluster/running-applications/job-submission/index.html
The pro of Ray Job submission is that it should be more robust than the other options. The con is that job submission is less interactive.

cc @richardliaw

**Comment by DmitriGekhtman** [comment]
@jbottum would it be fine if I attended the meeting at the above Zoom link tomorrow, along with a couple of other Ray and KubeRay maintainers?

**Comment by Jeffwan** [comment]
@juliusvonkohout 

> Enforce Runasnonroot for all images: Remove the root rights from the rayworker according to https://github.com/ray-project/ray/issues/14077#issuecomment-1283208595. I think udocker might be most promising. Most other images should already be compatible. @richardliaw can you take care of that?

I feel this is not a blocker? Seem that issue is to implement worker natively on Kubernetes.

**Comment by richardliaw** [comment]
Yeah the runasnonroot direction describes a different architectural direction that is not aligned with the recommended way of deploying Ray.

**Comment by jbottum** [comment]
@DmitriGekhtman please attend the Kubeflow Community Meeting and bring friends.    It would be great to organize and set expectations.

**Comment by juliusvonkohout** [comment]
> Yeah, +1 to @Jeffwan. the runasnonroot direction describes a different architectural direction that is not aligned with the recommended way of deploying Ray.

Which other deployment methods are there? I thought kuberay is the way to go.

Root containers are an obvious security blocker for official integration, so please tell me which Ray deployment method on Kubernetes works entirely with runasnonroot, such that we can integrate it.

There is usually no technical reason to use root.

**Comment by juliusvonkohout** [comment]
> @juliusvonkohout
> 
> Re: Configuration of RayCluster in UI. 
> 
> I believe this is what the KubeRay API server was built to support https://ray-project.github.io/kuberay/components/apiserver/. It's an option to use that here.
> cc @Jeffwan 
> 
> Re: Enforcing runasnonroot for workers.
> 
> Can you elaborate on this a bit more? Is this is an issue for the Ray team to solve? Can we modify or provide more options for how we build docker images?

The question is: why did you start with insecure images in the first place? I think that is on the ray team to solve, but I will try to help you. You can build the OCI images however you want, you just need to adhere to the official safe Kubernetes "Restricted PSP" standard https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted .

If you have a good architecture, just enforcing the restricted PSP on your namespace should be enough. You can also set the pod securitycontext manually to the values from the restricted set. The most important ones are to block anything that starts with host or privilege escalation and runasroot. If your images crash then please check whether you forgot to set proper file permissions on the working directories. If you need help on that level feel free to reach out on the Kubeflow slack or LinkedIn. I can also provide you Podsecuritypolicies if your clusters are below 1.23.

To properly build your images there are thousand of guides out there, but the most common stuff is :
For example use `USER 1000:0` at the end of the Dockerfiles to build the OCI container. Make sure working directories are created with 777 file permissions and of course do not use and drop all capabilities such as SYS_ADMIN, SYS_CHROOT (done by the restricted PSP) etc. Use proper networking ports above 1024, do not use insecure setuid or setuid binaries and so on, just the same stuff that you would do for any proper Linux userspace application.


This can of course happen in parallel with the other integration tasks.

**Comment by Jeffwan** [comment]
@juliusvonkohout 

> Can you elaborate on this a bit more? Is this is an issue for the Ray team to solve? Can we modify or provide more options for how we build docker images?

I agree that there're security concern. I think we can start with existing kuberay deployment. that means running workers inside the pod and managed by raylet. We can gradually improve the isolation level.

**Comment by jbottum** [comment]
Oct 25 Kubeflow Community Meeting update - We have created a "ray" slack channel on Kubeflow slack.   Dimitri and Julius are going to help lead the initial investigation, which might include a doc for architecture, and a user story (to help explain the user benefits).    Timing on the integration is TBD depending on scope, contributors and users.   @DmitriGekhtman @juliusvonkohout  I think we have to be careful about too many different touch points (i.e. this issue, Kubeflow (ray) slack channel, a google doc).   Can you please provide guidance on where people should comment and find current info ?

**Comment by juliusvonkohout** [comment]
> @juliusvonkohout re: insecure images, would you mind opening a [feature request issue on the Ray GitHub](https://github.com/ray-project/ray/issues/new?assignees=&labels=enhancement&template=feature-request.yml&title=%5B%3CRay+component%3A+Core%7CRLlib%7Cetc...%3E%5D+) ?

I have created https://github.com/ray-project/ray/issues/29665 for you. We really need an architecture meeting to estimate the whole integration.

**Comment by juliusvonkohout** [comment]
And this is the slack channel ID f anyone forgot to join C047L4B34KZ. It is within Kubeflows slack https://www.kubeflow.org/docs/about/community/

**Comment by juliusvonkohout** [comment]
> @juliusvonkohout
> 
> > Can you elaborate on this a bit more? Is this is an issue for the Ray team to solve? Can we modify or provide more options for how we build docker images?
> 
> I agree that there're security concern. I think we can start with existing kuberay deployment. that means running workers inside the pod and managed by raylet. We can gradually improve the isolation level.

As long as all pods (Worker/head/raylet) run with the restricted PSS that is fine. This is needed to prevent that the cluster can be more easily hacked as described in the kubernetes PSS documentation linked above. Adding isolation within the Worker/Raylet pod is not essential (but still desired in general) for kubeflow integration, since the users will only have access to their own on-demand created kuberay clusters in their own namespaces. They can damage their clusters in their own namespace anyway if they wish to do so.

Are there any architectural documents? We need an architectural proposal similar to https://docs.google.com/document/d/1DVHbT1RIv_VaIzWz77YCBYlBX7WEMW61AAm-ZHdWZFU/edit. 

I have created https://docs.google.com/document/d/1SP-COL6O-ETMQny7zXbrbEL4zZBG3REErWjfcvFZVEw/edit?usp=sharing&resourcekey=0-IaUGVemd2XX6MGTgifDbpg for that collaboration purpose.

**Comment by DmitriGekhtman** [comment]
> As long as all pods (Worker/head/raylet) run with the restricted PSS that is fine.

Sounds good.

General architecture description for KubeRay:
KubeRay pods are Ray nodes -- head and workers. Each Ray worker node pod runs several processes, including Ray _worker processes_, which run the actual workloads, and Raylets, which are local control processes.
The Ray head node pod is like the worker node pods, but it also has additional central control processes, including the GCS.

The reason for this many-concerns-in-a-pod set up is simply that Ray was not originally designed as a K8s-native application. The original set-up was to have the "Ray node" as the lowest level user-facing compute entity. So it is most convenient to match that up with the smallest stand-alone compute element in K8s (pod).

**Comment by Sharathmk99** [comment]
Hi All,

In our team we did integrate Ray cluster(kuberay) with Kubeflow and using it same on production cluster. 
We extended CRUD app similar to tensorboard for Ray Cluster as well, but we are yet to develop creation of ray cluster from UI. Currently user can create Ray cluster(Ray CRD, Virtual service and other CRD's) part of KF pipeline. User can delete existing ray cluster and also connect to Ray dashboard. See screenshot below,

![image](https://user-images.githubusercontent.com/3970340/198634029-b309680f-f776-49b1-ae1c-ec55409d7566.png)

We have not focused too much on security as we run our K8S cluster on-premise.

**Comment by juliusvonkohout** [comment]
> Hi All,
> 
> In our team we did integrate Ray cluster(kuberay) with Kubeflow and using it same on production cluster. We extended CRUD app similar to tensorboard for Ray Cluster as well, but we are yet to develop creation of ray cluster from UI. Currently user can create Ray cluster(Ray CRD, Virtual service and other CRD's) part of KF pipeline. User can delete existing ray cluster and also connect to Ray dashboard from UI, see screenshot below,
> 
> ![image](https://user-images.githubusercontent.com/3970340/198634029-b309680f-f776-49b1-ae1c-ec55409d7566.png)
> 
> We have not focused too much on security as we run our K8S cluster on-premise.

running on premises without an air gap means that you should be even more careful because most likely your Kubernetes team cannot compete with AWS and GCP ;-)

By the way what you say is quite confusing, because premise means something completely different than premises.
- Premise:  a proposition supporting or helping to support a conclusion.
- Premises:  a tract of land including its buildings."

Nevertheless this could be quite useful for @PierrickPochelu. 

1. Are you willing to add a the creation of ray clusters in the webinterface? 
2. Do you have an open source repository such that others can contribute? 
3. Are you using the raykube apiserver or are you creating Kubernetes resources directly?

**Comment by Sharathmk99** [comment]
> By the way what you say is quite confusing, because premise means something completely different than premises.
> 
> * Premise:  a proposition supporting or helping to support a conclusion.
> * Premises:  a tract of land including its buildings.

Sorry for typo. I meant we are running Kubernetes on-premises.



> running on premises without an air gap means that you should be even more careful because most likely your Kubernetes team cannot compete with AWS and GCP ;-)

Our Kubernetes team manages the on-premises cluster and our customers are our company ML engineers.



> 1. Are you willing to add a the creation of ray clusters in the webinterface?

Yes adding UI to create ray cluster is in our roadmap. Currently no timeline decided.



> 2\. Do you have an open source repository such that others can contribute?

No we don't have open source repository for now. I need to check internally how to handle the same.

I hope my answer helps you. Thanks!!

> 3\. Are you using the raykube apiserver or are you creating Kubernetes resources directly?

We are create K8S resources directly, not using ray apiserver.

**Comment by juliusvonkohout** [comment]
> Yes adding UI to create ray cluster is in our roadmap. Currently no timeline decided.
> 
> > 2. Do you have an open source repository such that others can contribute?
> 
> No we don't have open source repository for now. I need to check internally how to handle the same.

Yes please, this would speed up our development here and lower your long-term maintenance burden.

**Comment by juliusvonkohout** [comment]
@Sharathmk99 are you on board with the architecture here https://docs.google.com/document/d/1SP-COL6O-ETMQny7zXbrbEL4zZBG3REErWjfcvFZVEw/edit?usp=sharing&resourcekey=0-IaUGVemd2XX6MGTgifDbpg ?

1. Are you employing the operator into ray-system or the kubeflow namespace?  
2. Did you keep the istio-sidecars enabled for head and worker?
3. Did you need to add any permissions to the user namespace to proxy the dashboard. 

Also for the others please add comments to the architecture document.

**Comment by juliusvonkohout** [comment]
Please test https://github.com/kubeflow/manifests/pull/2383
